File: src/lib/server/contracts.ts
import { z } from "zod";

import { GUIDED_READING_ID } from "@/lib/audio/reading-passage";

export const deviceMetadataSchema = z
  .object({
    inputA: z
      .object({ deviceId: z.string().max(256), label: z.string().max(256) })
      .strict(),
    inputB: z
      .object({ deviceId: z.string().max(256), label: z.string().max(256) })
      .strict(),
    requested: z.record(z.string(), z.unknown()).default({}),
    reported: z.record(z.string(), z.unknown()).default({}),
  })
  .strict();

export const createSessionSchema = z
  .object({
    referenceId: z.literal(GUIDED_READING_ID).default(GUIDED_READING_ID),
    devices: deviceMetadataSchema.optional(),
  })
  .strict();

export const uploadAuthorizationSchema = z
  .object({
    sessionId: z.string().uuid(),
    slot: z.enum(["A", "B"]),
    bytes: z
      .number()
      .int()
      .min(44)
      .max(4 * 1024 * 1024),
    sha256: z.string().regex(/^[a-f0-9]{64}$/),
  })
  .strict();

export const commitCaptureSchema = z
  .object({
    sessionId: z.string().uuid(),
    slot: z.enum(["A", "B"]),
    pathname: z
      .string()
      .regex(/^sessions\/[0-9a-f-]+\/captures\/[AB]-[12]\.wav$/),
    bytes: z
      .number()
      .int()
      .positive()
      .max(4 * 1024 * 1024),
    sha256: z.string().regex(/^[a-f0-9]{64}$/),
    durationMs: z.number().int().min(19_000).max(20_500),
    sampleRate: z.number().int().min(16_000).max(96_000),
    channels: z.literal(1),
    codec: z.enum(["pcm_s16le", "pcm_f32le"]).default("pcm_s16le"),
    metrics: z.record(z.string(), z.number().finite()).default({}),
  })
  .strict();

export const startUpgradeSchema = z
  .object({
    sessionId: z.string().uuid(),
    upgradeId: z.string().uuid(),
  })
  .strict();

export type CommitCaptureInput = z.infer<typeof commitCaptureSchema>;

const audioDescriptorSchema = z
  .object({
    sha256: z.string().regex(/^[a-f0-9]{64}$/),
    byte_size: z.number().int().positive(),
    codec: z.string(),
    sample_rate_hz: z.number().int().positive(),
    channels: z.literal(1),
    bits_per_sample: z.number().int().positive(),
    frame_count: z.number().int().positive(),
    duration_seconds: z.number().positive(),
  })
  .strict();

const metricsSchema = z
  .object({
    peak_dbfs: z.number(),
    rms_dbfs: z.number(),
    steady_noise_floor_dbfs: z.number(),
    crest_factor_db: z.number(),
    dynamics_range_db: z.number(),
    clipping_ratio: z.number(),
    dc_offset: z.number(),
    high_frequency_energy_ratio: z.number(),
    spectral_rolloff_hz: z.number(),
    compression_proxy: z.number(),
    reverb_proxy: z.number(),
  })
  .strict();

const captureAnalysisSchema = z
  .object({ audio: audioDescriptorSchema, metrics: metricsSchema })
  .strict();
const artifactReceiptSchema = z
  .object({
    object_path: z.string(),
    content_type: z.enum(["audio/wav", "application/json"]),
    byte_size: z.number().int().positive(),
    sha256: z.string().regex(/^[a-f0-9]{64}$/),
  })
  .strict();
const wavArtifactReceiptSchema = artifactReceiptSchema.extend({
  content_type: z.literal("audio/wav"),
});
const jsonArtifactReceiptSchema = artifactReceiptSchema.extend({
  content_type: z.literal("application/json"),
});

export const workerVersionSchema = z
  .object({
    api_schema: z.literal("1"),
    build_revision: z.string().regex(/^[a-f0-9]{40}$/),
    pipeline_revision: z.literal("signal-enhancer-audio/1.0.0"),
    dsp_revision: z.literal("restrained-dsp/1.0.0"),
    model_name: z.literal("resemble-enhance"),
    model_repository: z.literal("ResembleAI/resemble-enhance"),
    model_revision: z.literal("4e3510ce4a8391159f665903544c5150bee7b2cb"),
    model_checkpoint_sha256: z.literal(
      "f9d035f318de3e6d919bc70cf7ad7d32b4fe92ec5cbe0b30029a27f5db07d9d6",
    ),
    source_repository: z.literal(
      "https://github.com/resemble-ai/resemble-enhance",
    ),
    source_revision: z.literal("8e978149bfe8abab3eb77d965d579a111afdb0ff"),
    inference_profile: z.literal(
      "enhancer-stage2:nfe=32:solver=midpoint:lambd=0.35:tau=0.45",
    ),
  })
  .strict();

export type WorkerVersion = z.infer<typeof workerVersionSchema>;

export const workerResultSchema = z
  .object({
    schema_version: z.literal("1"),
    job_id: z.string().uuid(),
    attempt_id: z.string().uuid(),
    source: z.literal("A"),
    routing: z
      .object({
        requested_engine: z.literal("resemble"),
        used_engine: z.literal("resemble"),
        outcome: z.literal("enhanced"),
        fallback_code: z.null(),
      })
      .strict(),
    before: captureAnalysisSchema,
    after: captureAnalysisSchema,
    comparison_b: captureAnalysisSchema,
    artifacts: z
      .object({
        enhanced_wav: wavArtifactReceiptSchema,
        difference_json: jsonArtifactReceiptSchema,
        report_json: jsonArtifactReceiptSchema,
      })
      .strict(),
    versions: workerVersionSchema,
  })
  .strict();

File: worker/src/signal_enhancer_worker/contracts.py
"""Strict public API contracts.

Signed URLs are deliberately represented by ``SecretStr``. They must never be
included in response models, logs, or exception messages.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    SecretStr,
    StringConstraints,
    field_validator,
    model_validator,
)

ApiVersion = Literal["1"]
UuidString = Annotated[
    str,
    StringConstraints(
        strict=True,
        pattern=r"^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
    ),
]
Sha256 = Annotated[str, StringConstraints(strict=True, pattern=r"^[0-9a-f]{64}$")]
ObjectPath = Annotated[
    str,
    StringConstraints(
        strict=True,
        min_length=3,
        max_length=512,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9._/-]+$",
    ),
]
BoundedText = Annotated[str, StringConstraints(strict=True, min_length=1, max_length=240)]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, validate_default=True)


class ReferenceVersion(StrictModel):
    id: Literal["diagnostic-speech", "guided-reading-v1"]
    revision: Literal["v1", "1.0.0"]

    @model_validator(mode="after")
    def require_matching_protocol_revision(self) -> ReferenceVersion:
        supported_pairs = {
            ("diagnostic-speech", "v1"),
            ("guided-reading-v1", "1.0.0"),
        }
        if (self.id, self.revision) not in supported_pairs:
            raise ValueError("reference protocol and revision do not match")
        return self


class SignedGetObject(StrictModel):
    object_path: ObjectPath
    url: SecretStr
    # JSON has no native datetime type, so this field alone parses an RFC 3339 string.
    expires_at: Annotated[datetime, Field(strict=False)]
    byte_size: Annotated[int, Field(strict=True, ge=44, le=4 * 1024 * 1024)]
    sha256: Sha256
    content_type: Literal["audio/wav"]

    @field_validator("url")
    @classmethod
    def validate_url_length(cls, value: SecretStr) -> SecretStr:
        if not 1 <= len(value.get_secret_value()) <= 4096:
            raise ValueError("signed URL length is outside the permitted range")
        return value

    @field_validator("expires_at")
    @classmethod
    def require_timezone(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("expires_at must include a timezone")
        return value

    @field_validator("expires_at", mode="before")
    @classmethod
    def require_rfc3339_input(cls, value: object) -> object:
        if isinstance(value, datetime):
            return value
        if not isinstance(value, str) or "T" not in value or len(value) > 40:
            raise ValueError("expires_at must be an RFC 3339 string")
        return value

    @field_validator("object_path")
    @classmethod
    def validate_object_path(cls, value: str) -> str:
        _validate_path_segments(value)
        return value


class SignedPutObject(StrictModel):
    object_path: ObjectPath
    url: SecretStr
    expires_at: Annotated[datetime, Field(strict=False)]
    max_bytes: Annotated[int, Field(strict=True, ge=2, le=4 * 1024 * 1024)]
    content_type: Literal["audio/wav", "application/json"]

    @field_validator("url")
    @classmethod
    def validate_url_length(cls, value: SecretStr) -> SecretStr:
        if not 1 <= len(value.get_secret_value()) <= 4096:
            raise ValueError("signed URL length is outside the permitted range")
        return value

    @field_validator("expires_at")
    @classmethod
    def require_timezone(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("expires_at must include a timezone")
        return value

    @field_validator("expires_at", mode="before")
    @classmethod
    def require_rfc3339_input(cls, value: object) -> object:
        if isinstance(value, datetime):
            return value
        if not isinstance(value, str) or "T" not in value or len(value) > 40:
            raise ValueError("expires_at must be an RFC 3339 string")
        return value

    @field_validator("object_path")
    @classmethod
    def validate_object_path(cls, value: str) -> str:
        _validate_path_segments(value)
        return value


def _validate_path_segments(value: str) -> None:
    if "//" in value or "\\" in value:
        raise ValueError("object path is not canonical")
    if any(part in {"", ".", ".."} for part in value.split("/")):
        raise ValueError("object path contains an invalid segment")


class CaptureInputs(StrictModel):
    a: SignedGetObject
    b: SignedGetObject

    @model_validator(mode="after")
    def distinct_inputs(self) -> CaptureInputs:
        if self.a.object_path == self.b.object_path:
            raise ValueError("input object paths must be distinct")
        return self


class OutputDestinations(StrictModel):
    enhanced_wav: SignedPutObject
    difference_json: SignedPutObject
    report_json: SignedPutObject

    @model_validator(mode="after")
    def validate_outputs(self) -> OutputDestinations:
        expected = {
            "enhanced_wav": (self.enhanced_wav, "audio/wav", "enhanced.wav"),
            "difference_json": (self.difference_json, "application/json", "difference.json"),
            "report_json": (self.report_json, "application/json", "report.json"),
        }
        paths: set[str] = set()
        for item, content_type, suffix in expected.values():
            if item.content_type != content_type:
                raise ValueError("output content type does not match its artifact kind")
            if not item.object_path.endswith(f"/{suffix}"):
                raise ValueError("output object path does not match its artifact kind")
            if item.object_path in paths:
                raise ValueError("output object paths must be distinct")
            paths.add(item.object_path)
        return self


class BaseJobRequest(StrictModel):
    schema_version: ApiVersion
    job_id: UuidString
    attempt_id: UuidString
    reference: ReferenceVersion
    inputs: CaptureInputs


class AnalyzeRequest(BaseJobRequest):
    pass


class EnhanceRequest(BaseJobRequest):
    source: Literal["A"]
    outputs: OutputDestinations

    @model_validator(mode="after")
    def validate_attempt_scoping(self) -> EnhanceRequest:
        input_paths = {self.inputs.a.object_path, self.inputs.b.object_path}
        job_segment = self.job_id
        attempt_segment = self.attempt_id
        for output in (
            self.outputs.enhanced_wav,
            self.outputs.difference_json,
            self.outputs.report_json,
        ):
            parts = output.object_path.split("/")
            if job_segment not in parts or attempt_segment not in parts:
                raise ValueError("output paths must be scoped to the job and attempt")
            if output.object_path in input_paths:
                raise ValueError("an output path cannot overwrite an input")
        return self


class AudioMetadata(StrictModel):
    sha256: Sha256
    byte_size: int
    codec: Literal["pcm_s16le", "pcm_s24le", "pcm_s32le", "float32le"]
    sample_rate_hz: int
    channels: Literal[1]
    bits_per_sample: Literal[16, 24, 32]
    frame_count: int
    duration_seconds: float


class AnalysisMetrics(StrictModel):
    peak_dbfs: float
    rms_dbfs: float
    steady_noise_floor_dbfs: float
    crest_factor_db: float
    dynamics_range_db: float
    clipping_ratio: float
    dc_offset: float
    high_frequency_energy_ratio: float
    spectral_rolloff_hz: float
    compression_proxy: float
    reverb_proxy: float


class CaptureAnalysis(StrictModel):
    audio: AudioMetadata
    metrics: AnalysisMetrics


class MetricDelta(StrictModel):
    noise_floor_db: float
    dynamics_range_db: float
    high_frequency_energy_ratio: float
    clipping_ratio: float


class VersionInfo(StrictModel):
    api_schema: ApiVersion
    build_revision: str
    pipeline_revision: str
    dsp_revision: str
    model_name: Literal["none", "resemble-enhance"]
    model_revision: str | None


class AnalyzeResponse(StrictModel):
    schema_version: ApiVersion
    job_id: UuidString
    attempt_id: UuidString
    inputs: dict[Literal["a", "b"], CaptureAnalysis]
    delta_b_minus_a: MetricDelta
    observations: Annotated[list[BoundedText], Field(max_length=8)]
    versions: VersionInfo


class ArtifactReceipt(StrictModel):
    object_path: ObjectPath
    content_type: Literal["audio/wav", "application/json"]
    byte_size: int
    sha256: Sha256


class RoutingDecision(StrictModel):
    requested_engine: Literal["dsp", "resemble"]
    used_engine: Literal["dsp", "resemble"]
    outcome: Literal["enhanced", "dsp_fallback"]
    fallback_code: Literal["model_unavailable"] | None = None


class EnhanceResponse(StrictModel):
    schema_version: ApiVersion
    job_id: UuidString
    attempt_id: UuidString
    source: Literal["A"]
    routing: RoutingDecision
    before: CaptureAnalysis
    after: CaptureAnalysis
    comparison_b: CaptureAnalysis
    artifacts: dict[Literal["enhanced_wav", "difference_json", "report_json"], ArtifactReceipt]
    versions: VersionInfo


class ProductStage(StrEnum):
    receiving_capture = "receiving_capture"
    inspecting_signal = "inspecting_signal"
    detecting_noise_and_compression = "detecting_noise_and_compression"
    restoring_detail = "restoring_detail"
    polishing_dynamics = "polishing_dynamics"
    generating_difference_map = "generating_difference_map"
    preparing_report = "preparing_report"


class StageEvent(StrictModel):
    schema_version: ApiVersion
    type: Literal["stage"]
    job_id: UuidString
    attempt_id: UuidString
    sequence: Annotated[int, Field(ge=1, le=7)]
    stage: ProductStage


class ResultEvent(StrictModel):
    schema_version: ApiVersion
    type: Literal["result"]
    sequence: Literal[8]
    result: EnhanceResponse


class ErrorResponse(StrictModel):
    schema_version: ApiVersion = "1"
    request_id: str
    code: str
    message: str
    retryable: bool


class HealthResponse(StrictModel):
    status: Literal["ready", "starting"]
